File: src/main/java/com/extendedae_plus/integration/recipeViewer/emi/EmiRecipeAdaptable.java
package com.extendedae_plus.integration.recipeViewer.emi;

import dev.emi.emi.api.recipe.EmiRecipe;
import dev.emi.emi.api.recipe.EmiRecipeCategory;
import dev.emi.emi.api.stack.EmiIngredient;
import dev.emi.emi.api.stack.EmiStack;
import dev.emi.emi.api.widget.WidgetHolder;
import dev.emi.emi.bom.MaterialNode;
import dev.emi.emi.jemi.JemiRecipe;
import lombok.AccessLevel;
import lombok.Setter;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.crafting.RecipeHolder;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class EmiRecipeAdaptable implements EmiRecipe {
    private final EmiRecipe recipe;
    @Setter(AccessLevel.PACKAGE)
    private List<EmiIngredient> inputs;

    public EmiRecipeAdaptable(EmiRecipe recipe) {
        if (recipe instanceof EmiRecipeAdaptable)
            throw new IllegalArgumentException("Cannot wrap a wrapped EmiRecipe");
        this.recipe = recipe;
    }

    public static EmiRecipeAdaptable of(EmiRecipe recipe, MaterialNode nodeParent, long batches) {
        var adaptable = new EmiRecipeAdaptable(recipe);
        adaptable.adapt(nodeParent, batches);
        return adaptable;
    }

    public static EmiRecipe unbox(EmiRecipe boxed) {
        if (boxed instanceof EmiRecipeAdaptable adaptable)
            return adaptable.recipe;
        else return boxed;
    }

    public static @Nullable JemiRecipe<?> unboxJemi(Object boxed) {
        if (boxed instanceof JemiRecipe<?> recipeJemi)
            return recipeJemi;
        else if (boxed instanceof EmiRecipeAdaptable adaptable)
            return unboxJemi(adaptable.recipe);
        else
            return null;
    }

    public void adapt(MaterialNode nodeParent, long batches) {
        var mapped = HelperBoMRecipes.mapChildren(nodeParent, batches);
        HelperBoMRecipes.applyMappings(this, this.recipe, mapped);
        this.getOutputs().forEach(output -> output.setAmount(output.getAmount() * batches));
    }

    @Override
    public List<EmiIngredient> getInputs() {
        return this.inputs == null
                ? this.recipe.getInputs()
                : this.inputs;
    }

    @Override
    public List<EmiStack> getOutputs() {
        return this.recipe.getOutputs();
    }

    @Override
    public EmiRecipeCategory getCategory() {
        return this.recipe.getCategory();
    }

    @Override
    public @Nullable ResourceLocation getId() {
        return this.recipe.getId();
    }

    @Override
    public List<EmiIngredient> getCatalysts() {
        return this.recipe.getCatalysts();
    }

    @Override
    public int getDisplayWidth() {
        return this.recipe.getDisplayWidth();
    }

    @Override
    public int getDisplayHeight() {
        return this.recipe.getDisplayHeight();
    }

    @Override
    public void addWidgets(WidgetHolder widgets) {
        this.recipe.addWidgets(widgets);
    }

    @Override
    public boolean supportsRecipeTree() {
        return this.recipe.supportsRecipeTree();
    }

    @Override
    public boolean hideCraftable() {
        return this.recipe.hideCraftable();
    }

    @Override
    public @Nullable RecipeHolder<?> getBackingRecipe() {
        return this.recipe.getBackingRecipe();
    }
}

File: src/main/kotlin/com/extendedae_plus/integration/impl/recipeViewer/emi/EmiRecipeAdaptable.kt
package com.extendedae_plus.integration.impl.recipeViewer.emi

import dev.emi.emi.api.recipe.EmiRecipe
import dev.emi.emi.api.recipe.EmiRecipeCategory
import dev.emi.emi.api.stack.EmiIngredient
import dev.emi.emi.api.stack.EmiStack
import dev.emi.emi.api.widget.WidgetHolder
import dev.emi.emi.bom.MaterialNode
import dev.emi.emi.jemi.JemiRecipe

class EmiRecipeAdaptable(recipe: EmiRecipe) : EmiRecipe {
    private val recipe: EmiRecipe

    internal var inputs: List<EmiIngredient>? = null

    init {
        require(recipe !is EmiRecipeAdaptable) { "Cannot wrap a wrapped EmiRecipe" }
        this.recipe = recipe
    }

    fun adapt(nodeParent: MaterialNode, batches: Long) {
        val mapped = HelperBoMRecipes.mapChildren(nodeParent, batches)
        HelperBoMRecipes.applyMappings(this, this.recipe, mapped)
        this.outputs.forEach { it.amount *= batches }
    }

    override fun getInputs(): List<EmiIngredient> = this.inputs ?: this.recipe.inputs

    override fun getOutputs(): List<EmiStack> = this.recipe.outputs

    override fun getCategory(): EmiRecipeCategory = this.recipe.category

    override fun getId() = this.recipe.id

    override fun getCatalysts(): List<EmiIngredient> = this.recipe.catalysts

    override fun getDisplayWidth() = this.recipe.displayWidth

    override fun getDisplayHeight() = this.recipe.displayHeight

    override fun addWidgets(widgets: WidgetHolder) =
        this.recipe.addWidgets(widgets)

    override fun supportsRecipeTree() = this.recipe.supportsRecipeTree()

    override fun hideCraftable() = this.recipe.hideCraftable()

    override fun getBackingRecipe() = this.recipe.backingRecipe

    companion object {
        @JvmStatic
        fun of(recipe: EmiRecipe, nodeParent: MaterialNode, batches: Long): EmiRecipeAdaptable {
            val adaptable = EmiRecipeAdaptable(recipe)
            adaptable.adapt(nodeParent, batches)
            return adaptable
        }

        @JvmStatic
        fun EmiRecipe.unbox() = (this as? EmiRecipeAdaptable)?.recipe ?: this

        @JvmStatic
        fun unboxJemi(boxed: Any): JemiRecipe<*>? = when (boxed) {
            is JemiRecipe<*> -> boxed
            is EmiRecipeAdaptable -> unboxJemi(boxed.recipe)
            else -> null
        }
    }
}
